### ISSUE TITLE
unarchive: tilde not parsed in dest

### ISSUE THREAD

**Issue by JoePricey** [ansible/ansible-modules-core#348]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.7.2
##### Environment:

Debian Sid
##### Summary:

Using a ~ in the dest parameter to unarchive doesn't work.
##### Steps To Reproduce:

Run a playbook containing this:

```
- name: unarchive
  unarchive: copy=no src=~ldm/ldm.tar.gz dest=~ldm/ldm/src
```

With an archive at ~ldm/ldm.tar.gz and the directory ~ldm/ldm/src existing.
##### Expected Results:

```
TASK: [ldm | unarchive] ******************************************************* 
changed: [ldm1]
```
##### Actual Results:

```
TASK: [ldm | unarchive] ******************************************************* 
failed: [ldm1] => {"failed": true}
msg: dest '~ldm/ldm/src' must be an existing dir

FATAL: all hosts have already failed -- aborting
```

**Comment by abadger** [comment]
I haven't been able to reproduce yet and the code looks like it would treat tilde's correctly.

Can you verify that  this works:

```
- name: unarchive
  unarchive: copy=no src=/home/ldm/ldm.tar.gz dest=/home/ldm/ldm/src
```

**Comment by cdlm** [comment]
Here it looks like `~/somewhere` is expanded in the context of the host (Mac-like `/Users/me/somewhere`) rather than the guest (Ubuntu, e.g `/home/vagrant/somewhere`)

**Comment by JoePricey** [comment]
@abadger yep that works fine.

@cdlm 's idea idea looks intriguing... a broken assumption when copy=no is set.

**Comment by abadger** [comment]
Ah, @cdim's suggestion would explain things.  In 1.8.0 I fixed some action_plugins that were expanding tilde locally but should have been expanding it remotely instead. So if you're using 1.7.x the code code attempt to expand the tilde locally.

Note -- there were some problems with the code doing having to do with delegate_to and also inventory variables.  If you use 1.8.2, the only thing that shouldn't work is trying to use templated values in delegate_to:.  That's been fixed in devel if you need a patch for that.

Also note, @cdim's suggestion makes sense for the original report if ~ldm would expand differently on the remote and local system (or not at all on the local system).  If it expands the same both locally and remotely it would definitely be worthwhile to try the newer version but it may turn out to not be the problem.

**Comment by abadger** [comment]
I'll close this sometime this week as fixed in 1.8.x unless someone discovers that it's still broken there.

**Comment by abadger** [comment]
The initial commit fixing this is here:
https://github.com/ansible/ansible/commit/bc4272d2a26e47418c7d588208482d05a34a34cd

There were some subsequent commits fixing bugs in the initial implementation.  Most prominently this one:
https://github.com/ansible/ansible/commit/565e5bbdfc42afe29c4f53b303352d7e8406dcba

# Closing This Ticket

Hi!
We believe recent commits (likely detailed above) should resolve this question or problem for you.

This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
